File: src/getCoatNoise2.py
def getCoatNoise(f, wl, wBeam, T, materialParams, materialSub, materialLayer
        , dOpt, dcdp):
    """
    Inputs
    ------
    f: Frequency
    wl: Wavelength
    wBeam: Beam width
    T: Temperature
    materialParams: Material parameters (Python list of class Material)
    materialSub: Substrate Material (string)
    materialLayer:
    dOpt:
    dcdp:

    Outputs
    -------
    SbrZ
    StoZ
    SteZ
    StrZ
    BrLayer
    """

    # Reciprocal thermodynamic beta
    kT = 1.3806e-23 * T

    # Angular Frequency
    omega = 2 * pi * f

    # Substrate thermal, optical, and mechanical properties
    alphaSub = materialParams[materialSub].alpha
    cSub = materialParams[materialSub].C
    kappaSub = materialParams[materialSub].kappa
    ySub = materialParams[materialSub].Y
    pratSub = materialParams[materialSub].prat

    # Material property vectors
    nN = np.zeros(np.size(dOpt), np.float64)
    aN = np.zeros(np.size(dOpt), np.int32)
    alphaN = np.zeros(np.size(dOpt), np.float64)
    betaN = np.zeros(np.size(dOpt), np.float64)
    kappaN = np.zeros(np.size(dOpt), np.float64)
    cN = np.zeros(np.size(dOpt), np.float64)
    yN = np.zeros(np.size(dOpt), np.float64)
    pratN = np.zeros(np.size(dOpt), np.float64)
    phiN = np.zeros(np.size(dOpt), np.float64)
    for i in range(np.size(materialLayer)):
        nN[i] = materialParams[materialLayer[i]].n
        aN[i] = materialParams[materialLayer[i]].a
        alphaN[i] = materialParams[materialLayer[i]].alpha
        betaN[i] = materialParams[materialLayer[i]].beta
        kappaN[i] = materialParams[materialLayer[i]].kappa
        cN[i] = materialParams[materialLayer[i]].C
        yN[i] = materialParams[materialLayer[i]].Y
        pratN[i] = materialParams[materialLayer[i]].prat
        phiN[i] = materialParams[materialLayer[i]].phiM

    # Geometrical thicknesses
    dGeo = lambda * dOpt / nN
    dCoat = np.sum(dGeo)

    ##################
    # Brownian Noise #
    ##################

    # Layer contributions
    a1 = (1 + nN * dcdp / 2) ** 2 * (ySub / yN)
    a2 = (1 - pratSub - 2 * pratSub.^2)**2 * yN
    a3 = (1 + pratN) ** 2 * (1 - 2 * pratN) * ySub
    a = (1 + nN * dcdp / 2) ** 2 * (ySub / yN) +  a2 / a3
    b = 1 - pratN
    c = 1 - pratN - 2 * pratN.^2
    d = 1 - pratSub - 2 * pratSub.^2
    brLayer = a / b * c / d

    # Sum contributions
    a = 4 * kBT / (pi * wBeam^2 * w)
    b = 1 - pratSub - 2 * pratSub ** 2
    SbrZ = a * sum(dGeo * brLayer * phiN * b / ySub);

    ######################
    # Thermo-optic Noise #
    ######################

    # Substrate alpha
    alphaBarSub

    # Layer contributions

    # Thermo-elastic

    # Thermo-refractive

    # Total thermo-optic

    # Return
    SbrZ = 0
    StoZ = 0
    SteZ = 0
    StrZ = 0
    BrLayer = 0
    return SbrZ, StoZ, SteZ, StrZ, BrLayer

File: matlab/getCoatNoise2.m
% dOpt   = optical thickness / lambda of each layer
%        = geometrical thickness * refractive index / lambda


function [SbrZ, StoZ, SteZ, StrZ, brLayer] = getCoatNoise2(f, lambda, wBeam, Temp, ...
  materialParams, materialSub, materialLayer, dOpt, dcdp)

  % Boltzmann const and temperature
  kBT = 1.3807e-23 * Temp;
  
  % angular frequency
  w = 2 * pi * f;
    
  % substrate properties
  alphaSub = materialParams(materialSub).alpha;
  cSub = materialParams(materialSub).C;
  kappaSub = materialParams(materialSub).kappa;
  ySub = materialParams(materialSub).Y;
  pratSub = materialParams(materialSub).prat;
  
  % make vectors of material properties
  nN = zeros(size(dOpt));
  aN = zeros(size(dOpt));
  alphaN = zeros(size(dOpt));
  betaN = zeros(size(dOpt));
  kappaN = zeros(size(dOpt));
  cN = zeros(size(dOpt));
  yN = zeros(size(dOpt));
  pratN = zeros(size(dOpt));
  phiN = zeros(size(dOpt));
  
  for n = 1:numel(materialLayer)
    nN(n) = materialParams(materialLayer(n)).n;
    aN(n) = materialParams(materialLayer(n)).a;
    alphaN(n) = materialParams(materialLayer(n)).alpha;
    betaN(n) = materialParams(materialLayer(n)).beta;
    kappaN(n) = materialParams(materialLayer(n)).kappa;
    cN(n) = materialParams(materialLayer(n)).C;
    yN(n) = materialParams(materialLayer(n)).Y;
    pratN(n) = materialParams(materialLayer(n)).prat;
    phiN(n) = materialParams(materialLayer(n)).phiM;
  end

  % geometrica thickness of each layer and total
  dGeo = lambda * dOpt ./ nN;
  dCoat = sum(dGeo);
  
  %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
  % Brownian
  %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
  % layer contrubutions
  brLayer = ...
    ( (1 + nN .* dcdp / 2).^2 .* (ySub ./ yN) + ...
     (1 - pratSub - 2 * pratSub.^2).^2 .* yN ./ ...
     ((1 + pratN).^2 .* (1 - 2 * pratN) .* ySub) )./ (1 - pratN) .* ((1 - pratN - 2 * pratN.^2))./((1 - pratSub - 2 * pratSub.^2));

  % sum them up for total
  SbrZ = (4 * kBT ./ (pi * wBeam^2 * w)) * ...
    sum(dGeo .* brLayer .* phiN * (1 - pratSub - 2 * pratSub.^2) / ySub);
  
  %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
  % Thermo-optic
  %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
  % substrate alpha
  alphaBarSub = 2 * (1 + pratSub) * alphaSub;

  % layer contrubutions
  alphaBar = (dGeo / dCoat) .* ((1 + pratSub) ./ (1 - pratN)) .* ...
    ((1 + pratN) / (1 + pratSub) + (1 - 2 * pratSub) .* yN / ySub) .* alphaN;
  
  betaBar = (-dcdp) .* dOpt .* ...
    (betaN ./ nN + alphaN .* (1 + pratN) ./ (1 - pratN));

  % thermo-elastic
  SteZ = (4 * kBT * Temp ./ (pi * wBeam^2 * sqrt(2 * kappaSub * cSub * w))) * ...
    ( sum(alphaBar * dCoat) - alphaBarSub * sum(dGeo .* cN) / cSub ).^2;
    
  % thermo-refractive
  StrZ = (4 * kBT * Temp ./ (pi * wBeam^2 * sqrt(2 * kappaSub * cSub * w))) * ...
    sum(betaBar * lambda).^2;
    
  % total thermo-optic
  StoZ = (4 * kBT * Temp ./ (pi * wBeam^2 * sqrt(2 * kappaSub * cSub * w))) * ...
    ( sum(alphaBar * dCoat) - sum(betaBar * lambda) - ...
      alphaBarSub * sum(dGeo .* cN) / cSub ).^2;
  
end
